File: PROGRAMMERS/LEVEL2/프린터.cpp
#include <string>
#include <queue>
#include <vector>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 1;
    //기본으로 우선순위 큐는 내림차순이다.
    priority_queue<int> pq;
    queue<pair<int, int>>q;
    for (int i = 0; i < priorities.size(); i++) {
        pq.push(priorities[i]);
        q.push(make_pair(i, priorities[i]));
    }
    while (!q.empty()) {
        //우선순위큐 top은 맨 앞 원소이다.
        if (q.front().second == pq.top()) {
            if (q.front().first == location)
                return answer;
            else {
                answer++;
                q.pop();
                pq.pop();
            }
        }
        else {
            q.push(q.front());
            q.pop();
        }
    }
    return answer;
}

File: PROGRAMMERS/LEVEL2/프린터.java
import java.util.*;
import java.util.Map.Entry;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 1;
        
        Queue<Integer> pq = new PriorityQueue<Integer>(Collections.reverseOrder());
        Queue<int[]> q = new LinkedList<int[]>();
        
        for(int i=0;i<priorities.length;i++) {
        	pq.add(priorities[i]);
        	q.add(new int[] {i,priorities[i]});
        }
        
        while(!q.isEmpty()) {
        	int[] now = q.poll();
        	if(now[1]==pq.peek()) {
        		if(now[0]==location) {
        			return answer;
        		}
        		else {
        			answer++;
        			pq.poll();
        		}
        	}
        	else {
        		q.add(now);
        	}
        }
        
        
        
        return answer;
    }
}